File: commands/config/prefix.go
package config

import (
  "github.com/sapphire-cord/sapphire"
  "github.com/pollen5/taiga/db"
)

// TODO: permissions

// Sets the prefix for this server.
// Usage: [prefix:string]
// Aliases: setprefix, changeprefix
// guild only
func Prefix(ctx *sapphire.CommandContext) {
  // If no arguments, show the existing prefix.
  // e.g for lost people to use @botname prefix
  if !ctx.HasArgs() {
    ctx.Reply("The current prefix for this server is `%s`", db.GetPrefix(ctx.Guild.ID))
    return
  }

  prefix := ctx.Arg(0).AsString()

  // Set a soft limit to prevent spam.
  if len(prefix) > 10 {
    ctx.Reply("Prefix must not be longer than 10 characters.")
    return
  }

  _, err := db.Exec("INSERT INTO guilds (id, prefix) VALUES ($1, $2) ON CONFLICT (id) DO UPDATE SET prefix=$2", ctx.Guild.ID, prefix)
  if err != nil {
    ctx.Error(err)
    return
  }

  ctx.Reply("Successfully updated the prefix for this server to `%s`", prefix)
}

File: commands/config/prefix.js
const Command = require("../../structures/Command.js");

class Prefix extends Command {
  constructor(...args) {
    super(...args, {
      description: "Set or reset the prefix for this server.",
      usage: "prefix [prefix|reset]",
      guildOnly: true,
      aliases: ["setprefix", "changeprefix"]
    });
  }
  
  async run(ctx, args) {
    if(!args.length) {
      return ctx.reply(`The prefix for this server is \`${ctx.guild.settings.prefix}\``);
    }

    if(!ctx.member.permissions.has("MANAGE_GUILD"))
      return ctx.reply("Baka! You need the `Manage Server` permissions to change the prerix.");

    const prefix = args.join(" ");

    if(prefix === "reset") return this.reset(ctx);
    if(prefix.length > 10) return ctx.reply("Prefix can't be longer than 10 characters.");
    if(prefix === ctx.guild.settings.prefix) throw "Baka! That is already the current prefix.";

    await ctx.guild.update({ prefix });
    return ctx.reply(`Successfully updated prefix to: \`${prefix}\``);
  }
  
  async reset(ctx) {
    if(ctx.guild.settings.prefix === "m!") return ctx.reply("The prefix is already the default.");
    await ctx.guild.update({ prefix: "m!" })
    return ctx.reply("Reset the prefix for this server to `m!`");
  }
}

module.exports = Prefix;
